docs/reference/wasm.md

```diff
@@ -0,0 +1,242 @@
+---
+title: 'wasm'
+weight: 110
+description: 'Wasm API exposed to guests'
+---
+
+Inspektor Gadget exposes some functions to wasm modules implemented in gadgets.
+We provide a Golang wrapper for this functionality, but these functions can be
+used directly from any programming language that can be compiled to wasm.
+
+## Data types
+
+Data types passed to the API are encoded using 64 bits. Primitive types like
+integers, booleans and floats are casted directly to 64 bits integer and passed
+using the stack.
+
+Strings and byte arrays are stored in the wasm module's memory. A 64 bits
+integer is used to represent a pointer to them, the higher 32 bits contains the
+length and the lower 32 the memory address.
+
+## API
+
+### Log
+
+####  `gadgetLog(u32 level, string msg)`
+
+Print a log message using the gadget's logger instance.
+
+Parameters:
+
+- `level` (u32): Log level:
+  - 0: Error
+  - 1: Warn
+  - 2: Info
+  - 3: Debug
+  - 4: Trace
+- `msg` (string): Message to print
+
+Return value:
+- None
+
+### Datasources
+
+#### `newDataSource(string name)`
+
+Creates a new data source.
+
+Parameters:
+- `name` (string): Data source name
+
+Return value:
+- (u32): Handle to the created data source on success, 0 on error.
+
+#### `getDataSource(string name) u32`
+
+Get a handle to a data source.
+
+Parameters:
+- `name` (string): Datasource's name
+
+Return value:
+- (u32) Handle to the data source on success, 0 on error.
+
+#### `dataSourceSubscribe(u32 ds, u32 type, u32 prio, u64 cb)`
+
+Subscribe to events emitted by a data source.
+
+This mechanism requires the wasm module to export a `dsCallback` that is called
+by the host when an event is emitted:
+
+`dsCallback(u64 cbID, u32 ds, u32 data)`
+- `cbID`: Callback ID
+- `ds`: Datasource handle
+- `data`: Data handle
+
+
+Parameters:
+- `ds` (u32): Datasource handle (as returned by `getDataSource` or `newDataSource`)
+- `type` (u32): Subscription type: (0: Data, 1: Array, 2: Packet)
```

Comment:
What is `Packet`? A skb appended to a message in a BPF_MAP_TYPE_PERF_EVENT_ARRAY?
I think the 3 types should be explained in the documentation.

Reply:
A `Packet` is the type that is sent from the server to the client containing the events. I agree we need to document it, but IMO not on this PR but directly on the data source: 

https://github.com/inspektor-gadget/inspektor-gadget/blob/ba42e61eb89bc166ff6d5e9114b2277d37e8a371/pkg/datasource/datasource.go#L70-L76

I think we can handle it on the the "gadget developers" or "core concepts" part of #1184.